PhysicsTools/PatAlgos/plugins/PATTauProducer.cc

```diff
@@ -382,26 +383,46 @@ void PATTauProducer::produce(edm::Event & iEvent, const edm::EventSetup & iSetup
-     	else if ( icand->muonRef().isNonnull() && icand->muonRef()->globalTrack().isNonnull() ) track = icand->muonRef()->globalTrack().get();
-     	else if ( icand->muonRef().isNonnull() && icand->muonRef()->outerTrack().isNonnull()  ) track = icand->muonRef()->outerTrack().get();
-     	else if ( icand->gsfTrackRef().isNonnull() ) track = icand->gsfTrackRef().get();
-     	if( track ) {
-     	  if( track->pt() > leadChargedCandPt ) {
-     	    leadChargedCandEtaAtEcalEntrance = icand->positionAtECALEntrance().eta();
-     	    leadChargedCandPt = track->pt();
-     	  }
-        } 		
+      const std::vector<reco::CandidatePtr>& signalCands = pfTauRef->signalPFCands();
+      for(const auto& it : signalCands) {
+	const reco::PFCandidate* icand = dynamic_cast<const reco::PFCandidate*>(it.get());
+	if (icand != nullptr) {
+          ecalEnergy += icand->ecalEnergy();
+          hcalEnergy += icand->hcalEnergy();
+	  sumPhiTimesEnergy += icand->positionAtECALEntrance().phi()*icand->energy();		
+	  sumEtaTimesEnergy += icand->positionAtECALEntrance().eta()*icand->energy();
+          sumEnergy += icand->energy();	 
+	  const reco::Track* track = nullptr;
+     	  if ( icand->trackRef().isNonnull() ) track = icand->trackRef().get();
+     	else if ( icand->muonRef().isNonnull() && icand->muonRef()->innerTrack().isNonnull()  ) track = icand->muonRef  ()->innerTrack().get();
+     	else if ( icand->muonRef().isNonnull() && icand->muonRef()->globalTrack().isNonnull() ) track = icand->muonRef  ()->globalTrack().get();
+     	else if ( icand->muonRef().isNonnull() && icand->muonRef()->outerTrack().isNonnull()  ) track = icand->muonRef  ()->outerTrack().get();
+     	  else if ( icand->gsfTrackRef().isNonnull() ) track = icand->gsfTrackRef().get();
+     	  if( track ) {
+     	    if( track->pt() > leadChargedCandPt ) {
+     	      leadChargedCandEtaAtEcalEntrance = icand->positionAtECALEntrance().eta();
+     	      leadChargedCandPt = track->pt();
+     	    }
+          }
+        }
+        else{
+	  const pat::PackedCandidate* packedCandPtr = dynamic_cast<const pat::PackedCandidate*>(it.get());
+	  if(packedCandPtr != nullptr) {
+	    //position at ECAL entrance can be recomputed, but individual ECAL and HCAL energies not
+	    //ecalEnergy += packedCandPtr->ecalEnergy();
+	    //hcalEnergy += packedCandPtr->hcalEnergy();
+	    //sumPhiTimesEnergy += packedCandPtr->positionAtECALEntrance().phi()*packedCandPtr->energy();		
+	    //sumEtaTimesEnergy += packedCandPtr->positionAtECALEntrance().eta()*packedCandPtr->energy();
+	    sumEnergy += packedCandPtr->energy();
+	    const reco::Track* track = packedCandPtr->hasTrackDetails() ? &packedCandPtr->pseudoTrack() : nullptr;
+	    if( track != nullptr ) {
+	      if( track->pt() > leadChargedCandPt ) {
+		//leadChargedCandEtaAtEcalEntrance = packedCandPtr->positionAtECALEntrance().eta();
```

Comment:
same here. If we are sure what we are doing is fine we should remove the old stuff. if it still keeps functionality it should be commented in. If there is a real reasons to have something commented a clear statement should be added that RECO can understand ;-)...